#! /bin/sh -e

# All lines beginning with `# DP:' are a description of the patch.
# DP: Description: Fix glibc build breakage with binutils 2.15.
# DP: Related bugs: #266598: glibc: FTBFS on sparc
# DP: Dpatch author: GOTO Masanori <gotom@debian.org>
# DP: Patch author: Jakub Jelinek <jakub@redhat.com>
# DP: Upstream status: In CVS
# DP: Status Details: already in cvs.
# DP: Date: 2004-10-05

PATCHLEVEL=0

if [ $# -ne 2 ]; then
    echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
    exit 1
fi
case "$1" in
    -patch) patch -d "$2" -f --no-backup-if-mismatch -p$PATCHLEVEL < $0;;
    -unpatch) patch -d "$2" -f --no-backup-if-mismatch -R -p$PATCHLEVEL < $0;;
    *)
	echo >&2 "`basename $0`: script expects -patch|-unpatch as argument"
	exit 1
esac
exit 0

# append the patch here and adjust the -p? flag in the patch calls.
2004-04-29  Jakub Jelinek  <jakub@redhat.com>

	* sysdeps/sparc/sparc64/soft-fp/qp_qtoi.c (_Qp_qtoi): Use %f31
	for single precision register, add it to __asm clobbers [BZ #139].
	* sysdeps/sparc/sparc64/soft-fp/qp_qtoui.c (_Qp_qtoui): Use %f31
	for single precision register, add it to __asm clobbers.
	* sysdeps/sparc/sparc64/soft-fp/qp_qtoux.c (_Qp_qtoux): Use fqtox
	instead of fqtoi in QP_HANDLE_EXCEPTIONS.
	* sysdeps/sparc/sparc64/soft-fp/qp_qtox.c (_Qp_qtox): Likewise.
	Reported by M. H. VanLeeuwen <vanl@megsinet.net>.

Index: sysdeps/sparc/sparc64/soft-fp/qp_qtoi.c
===================================================================
RCS file: /cvs/glibc/libc/sysdeps/sparc/sparc64/soft-fp/qp_qtoi.c,v
retrieving revision 1.3
retrieving revision 1.4
diff -u -r1.3 -r1.4
--- sysdeps/sparc/sparc64/soft-fp/qp_qtoi.c	6 Jul 2001 04:56:05 -0000	1.3
+++ sysdeps/sparc/sparc64/soft-fp/qp_qtoi.c	29 Apr 2004 20:02:34 -0000	1.4
@@ -1,6 +1,6 @@
 /* Software floating-point emulation.
    Return (int)(*a)
-   Copyright (C) 1997,1999 Free Software Foundation, Inc.
+   Copyright (C) 1997, 1999, 2004 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
    Contributed by Richard Henderson (rth@cygnus.com) and
 		  Jakub Jelinek (jj@ultra.linux.cz).
@@ -38,9 +38,9 @@
   	__asm (
 "	ldd [%1], %%f52\n"
 "	ldd [%1+8], %%f54\n"
-"	fqtoi %%f52, %%f60\n"
-"	st %%f60, [%0]\n"
-"	" : : "r" (&rx), "r" (a) : QP_CLOBBER);
+"	fqtoi %%f52, %%f31\n"
+"	st %%f31, [%0]\n"
+"	" : : "r" (&rx), "r" (a) : QP_CLOBBER, "f31");
 	r = rx);
 
   return r;
Index: sysdeps/sparc/sparc64/soft-fp/qp_qtoui.c
===================================================================
RCS file: /cvs/glibc/libc/sysdeps/sparc/sparc64/soft-fp/qp_qtoui.c,v
retrieving revision 1.3
retrieving revision 1.4
diff -u -r1.3 -r1.4
--- sysdeps/sparc/sparc64/soft-fp/qp_qtoui.c	6 Jul 2001 04:56:05 -0000	1.3
+++ sysdeps/sparc/sparc64/soft-fp/qp_qtoui.c	29 Apr 2004 20:02:34 -0000	1.4
@@ -1,6 +1,6 @@
 /* Software floating-point emulation.
    Return (unsigned int)(*a)
-   Copyright (C) 1997,1999 Free Software Foundation, Inc.
+   Copyright (C) 1997, 1999, 2004 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
    Contributed by Richard Henderson (rth@cygnus.com) and
 		  Jakub Jelinek (jj@ultra.linux.cz).
@@ -38,9 +38,9 @@
   	__asm (
 "	ldd [%1], %%f52\n"
 "	ldd [%1+8], %%f54\n"
-"	fqtoi %%f52, %%f60\n"
-"	st %%f60, [%0]\n"
-"	" : : "r" (&rx), "r" (a) : QP_CLOBBER);
+"	fqtoi %%f52, %%f31\n"
+"	st %%f31, [%0]\n"
+"	" : : "r" (&rx), "r" (a) : QP_CLOBBER, "f31");
 	r = rx);
 
   return r;
Index: sysdeps/sparc/sparc64/soft-fp/qp_qtoux.c
===================================================================
RCS file: /cvs/glibc/libc/sysdeps/sparc/sparc64/soft-fp/qp_qtoux.c,v
retrieving revision 1.3
retrieving revision 1.4
diff -u -r1.3 -r1.4
--- sysdeps/sparc/sparc64/soft-fp/qp_qtoux.c	6 Jul 2001 04:56:05 -0000	1.3
+++ sysdeps/sparc/sparc64/soft-fp/qp_qtoux.c	29 Apr 2004 20:02:34 -0000	1.4
@@ -1,6 +1,6 @@
 /* Software floating-point emulation.
    Return (unsigned long)(*a)
-   Copyright (C) 1997,1999 Free Software Foundation, Inc.
+   Copyright (C) 1997, 1999, 2004 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
    Contributed by Richard Henderson (rth@cygnus.com) and
 		  Jakub Jelinek (jj@ultra.linux.cz).
@@ -38,7 +38,7 @@
   	__asm (
 "	ldd [%1], %%f52\n"
 "	ldd [%1+8], %%f54\n"
-"	fqtoi %%f52, %%f60\n"
+"	fqtox %%f52, %%f60\n"
 "	std %%f60, [%0]\n"
 "	" : : "r" (&rx), "r" (a) : QP_CLOBBER);
 	r = rx);
Index: sysdeps/sparc/sparc64/soft-fp/qp_qtox.c
===================================================================
RCS file: /cvs/glibc/libc/sysdeps/sparc/sparc64/soft-fp/qp_qtox.c,v
retrieving revision 1.3
retrieving revision 1.4
diff -u -r1.3 -r1.4
--- sysdeps/sparc/sparc64/soft-fp/qp_qtox.c	6 Jul 2001 04:56:05 -0000	1.3
+++ sysdeps/sparc/sparc64/soft-fp/qp_qtox.c	29 Apr 2004 20:02:34 -0000	1.4
@@ -1,6 +1,6 @@
 /* Software floating-point emulation.
    Return (long)(*a)
-   Copyright (C) 1997,1999 Free Software Foundation, Inc.
+   Copyright (C) 1997, 1999, 2004 Free Software Foundation, Inc.
    This file is part of the GNU C Library.
    Contributed by Richard Henderson (rth@cygnus.com) and
 		  Jakub Jelinek (jj@ultra.linux.cz).
@@ -38,7 +38,7 @@
   	__asm (
 "	ldd [%1], %%f52\n"
 "	ldd [%1+8], %%f54\n"
-"	fqtoi %%f52, %%f60\n"
+"	fqtox %%f52, %%f60\n"
 "	std %%f60, [%0]\n"
 "	" : : "r" (&rx), "r" (a) : QP_CLOBBER);
 	r = rx);
